### PR TITLE
remove unnecessary waitOnEPStep when current node and the consumer node are in the same stream

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
Remove the unnecessary WaitOnEPStep if the current operator node and its consumer are in the same stream while there are notifications filed in the current node

### Motivation and Context
In the current code, the WaitOnEPStep will always be launched as long as the notification is filed in the input node, no matter the current node and the input node are in the same stream or not, which is not necessary.
This PR is to remove the WaitOnEPStep for this case.

### CODE REVIEW

File: onnxruntime/core/framework/allocation_planner.cc [hunk 1]
```diff
@@ -1781,11 +1812,12 @@ class PlannerImpl {
...
-    //    a. which node need to generate notification
-    //    b. which node need to trigger downstream
-    size_t num_trigger_points = 0;
-    InlinedHashMap<NodeIndex, size_t> node_to_trigger_points;
+    // 2. determing which nodes to generate notification and/or trigger downstream
+    // diverge_stream_from_node[i] = j means node i and its consumer nodes are splitted into different streams
+    // node_to_notification[i] = j means node i and its consumer nodes are in the same stream, but still need notification for some special cases
+    // j is the index to determine the corresponding notification owner in the following calculation.
+    size_t diverge_stream_node_count = 0;
+    InlinedHashMap<NodeIndex, size_t> diverge_stream_from_node;
```
Comment (souptc) [inline onnxruntime/core/framework/allocation_planner.cc:1820]: here seems we imply that the "diverge_stream_from_node" will always trigger a notification. But for the case like both producer and consumer are CPU nodes, there is no notification creator registered. In that case, I guess the notification will result in a nullptr. will we generate redundant ActivateNotification and Wait step for such a case?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 2]
```diff
@@ -1761,6 +1761,37 @@ class PlannerImpl {
     num_logic_streams_ = stream_nodes_.size();
   }
 
+  void ConstructWaitOnEPStep(const onnxruntime::Node* node, Node::NodeConstIterator it, size_t notification_owner_index,
```
[the remark this answers is not in the document]
Comment (souptc) [reply onnxruntime/core/framework/allocation_planner.cc:1764]: and consider make this function static.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 1 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1820]: If both producers and consumers are CPU nodes, then diverge_stream_from_node will be empty. Whether there will be ActivateNotification and wait step is determined by the map node_to_notification. So I don't think there will be redundant ActivateNotification and wait step for this case.

File: onnxruntime/core/framework/allocation_planner.cc [hunk 3]
```diff
@@ -1867,11 +1897,14 @@ class PlannerImpl {
         }
         auto* node = graph_viewer_.GetNode(node_index);
         for (auto it = node->InputNodesBegin(); it != node->InputNodesEnd(); ++it) {
-          //  check whether we need to add barrier
+          InlinedHashSet<OrtDevice::DeviceType> devices_to_wait_on;
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1900]: I think same issue also exist on the current code?

File: onnxruntime/core/framework/allocation_planner.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1764]: comments added. We cannot make it static as it uses some instance variables like plan_

File: onnxruntime/core/framework/allocation_planner.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (souptc) [reply onnxruntime/core/framework/allocation_planner.cc:1905]: then let's fix it :)

File: onnxruntime/core/framework/allocation_planner.cc [hunk 4]
```diff
@@ -1761,6 +1761,44 @@ class PlannerImpl {
...
 
+  // Construct WaitOnEPStep if has EP registered it.
+  // <param name="node">Current node to inspect</param>
+  // <param name="producer_node">producer node of the current node</param>
+  // <param name="notification_owner_index">index to find the corresponding stream owns the notification</param>
+  // <param name="stream_handle_registry">stream handle registry to get the wait handler</param>
+  // <param name="execution_plan">execution plan</param>
+  // <param name="stream_index">index to get current logic stream</param>
+  // <param name="devices_to_wait_on">Cached set of the devices to wait on</param>
+  void ConstructWaitOnEPStep(const Node* node, const Node& producer_node, size_t notification_owner_index,
```
Comment (souptc) [inline onnxruntime/core/framework/allocation_planner.cc:1772]: static function?

File: onnxruntime/core/framework/allocation_planner.cc (hunk 1 above)
Comment (souptc) [reply onnxruntime/core/framework/allocation_planner.cc:1826]: what about if you have multiple streams (through manual configuration) on CPU nodes

File: onnxruntime/core/framework/allocation_planner.cc [hunk 5]
```diff
@@ -1794,50 +1832,44 @@
         for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it) {
           // if the output node is not in the same stream, generate a trigger point
           if (node_stream_map_[it->Index()] != i) {
-            node_to_trigger_points[node_index] = num_trigger_points;
-            num_trigger_points++;
+            diverge_stream_from_node[node_index] = num_notifications;
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1835]: diverge_stream_from_node[node_index] = num_notifications++;

File: onnxruntime/core/framework/allocation_planner.cc [hunk 6]
```diff
@@ -1794,50 +1832,44 @@
...
-                  if (node_stream_map_[it->Index()] != i ||
-                      (node_stream_map_[it->Index()] == i &&
+                  // the consumer is in the same stream(non-cpu device), but it consumes a CPU tensor from an non-shape op.
+                  // for example, a resize cuda kernel consumer a tensor from MemCpyToHost cuda kernel on the same stream.
+                  // in this case, the FIFO can't guarantee the cpu tensor is ready when resize kernel is launching
+                  if ((node_stream_map_[it->Index()] == i &&
                        execution_plan[i]->device_.Type() != OrtDevice::CPU &&
                        plan_.allocation_plan[output_arg_idx].location.device.Type() == OrtDevice::CPU)) {
-                    requires_notification = true;
+                    node_to_notification[node_index] = num_notifications;
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1860]: > node_to_notification[node_index] = num_notifications; [](http://example.com/codeflow?start=19&length=54)

"node_to_notification[node_index] = num_notifications++;"?
And we are deliberately incrementing  num_notifications both on line 1836 and here?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 7]
```diff
@@ -1848,12 +1880,15 @@
         ORT_ENFORCE(execution_plan[node_stream_map_[node_index]]->device_.Type() == node_device_mem_location.device.Type());
       }
     }
-    // 4. set notification owners
+    // 4. set notification owners. Notice that both diverge_stream_from_node and node_to_notification will contribute to notification_owners
     plan_.notification_owners.resize(num_notifications);
     for (auto node_index : graph_viewer_.GetNodesInTopologicalOrder(context_->GetExecutionOrder())) {
-      auto it = node_to_notification.find(node_index);
+      auto it = diverge_stream_from_node.find(node_index);
+      if (it != diverge_stream_from_node.end()) plan_.notification_owners[it->second] = node_stream_map_[node_index];
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1887]: >if (it != diverge_stream_from_node.end()) plan_.notification_owners[it->second] = node_stream_map_[node_index]; [](http://example.com/codeflow?start=6&length=111)

if (...) {
...
}

File: onnxruntime/core/framework/allocation_planner.cc [hunk 8]
```diff
@@ -1848,12 +1880,15 @@
...
+
+      it = node_to_notification.find(node_index);
       if (it != node_to_notification.end()) {
-        // notification owned by the node who produced it.
+        // notification owned by the stream who produced it.
         plan_.notification_owners[it->second] = node_stream_map_[node_index];
       }
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1893]: Assume "it" are both valid on 1887 and 1890, will it->second will be different or same?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 9]
```diff
@@ -1882,37 +1920,11 @@
...
-                    }
-                    devices_to_wait_on.insert(consumer_device.Type());
-                  }
-                }
-              }
-            }
+            ConstructWaitOnEPStep(node, *it, trigger_point_it->second, stream_handle_registry, execution_plan, i, devices_to_wait_on);
+          } else if ((notification_it = node_to_notification.find(it->Index())) != node_to_notification.end()) {
+            // Else the current node and its input node are in the same stream, but we can still find input node in node_to_notification,
+            // which means we will launch WaitOnEPStep only before LaunchKernelStep.
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1926]: For "else" what if we do not launch WaitOnEpStep? Still we have the right sequence no?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 10]
```diff
@@ -1882,37 +1920,11 @@
...
-#ifdef ENABLE_TRAINING
-                      execution_plan[i]->step_pc.push_back(node_index);
-#endif
-                    }
-                    devices_to_wait_on.insert(consumer_device.Type());
-                  }
-                }
-              }
-            }
+            ConstructWaitOnEPStep(node, *it, trigger_point_it->second, stream_handle_registry, execution_plan, i, devices_to_wait_on);
```
Comment (RandySheriffH) [inline onnxruntime/core/framework/allocation_planner.cc:1923]: For "if", there is an interesting case - suppose node C needs to wait on A and B, which are both on the same stream, should we insert two waits or simply one that enough?

File: onnxruntime/core/framework/allocation_planner.cc (hunk 4 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1772]: As I replied in your earlier comment, we cannot make it static as it uses some instance variables like plan_

File: onnxruntime/core/framework/allocation_planner.cc (hunk 1 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1826]: I don't quite follow your concern. diverge_stream_from_node only records the case that current node starting to diverge to different streams, no matter these streams are CPU nodes or GPU nodes. We will launch the corresponding ActivateNotificationStep for the nodes in this map (and WaitOnEPStep for its consumer nodes). For the nodes that are in the same stream, there are still chance to launch ActivateNotificationStep (and corresponding WaitOnEPStep) from node_to_notification

File: onnxruntime/core/framework/allocation_planner.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1905]: removed devices_to_wait_on. Just curious why it is introduced at the very beginning.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 6 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1860]: Yes. size(notification_owner) == size(diverge_stream_from_node) + size(node_to_notification). See the newly added comments, both diverge_stream_from_node and node_to_notification is capable of launching ActivateNotification step.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 8 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1893]: It's not possible. diverge_stream_from_node records the case that nodes and their consumers are in different streams, while node_to_notification records the case that nodes and their consumers are in the same stream.

Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1893]: I think it deserves a UT for this scenario: a node A is in different stream with one of its consumer node B (so a pair of ActivateNotification/wait steps will be needed), and in the same stream with another of its consumer node C which still needs to launch ActivateNotification. The previous code will miss one pair, I think this PR will handle this case well.

File: onnxruntime/core/framework/allocation_planner.cc (hunk 9 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1926]: If we do not launch WaitOnEpStep under "else if" clause, that means there exist a consumer node which is in the same stream as its producer node, and it should have launched WaitOnEpStep (for the corresponding tensors are in different device). We will not have the right sequence for this case

File: onnxruntime/core/framework/allocation_planner.cc (hunk 10 above)
Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1923]: I've removed the devices_to_wait_on cache. So in the latest commit of this PR we will have 2 waits

Comment (RandySheriffH) [reply onnxruntime/core/framework/allocation_planner.cc]: Exactly - shouldn't wait once be quite sufficient?

File: onnxruntime/core/framework/allocation_planner.cc [hunk 11]
```diff
@@ -1896,24 +1891,26 @@ class PlannerImpl {
...
                 if (std::find(it->OutputDefs().begin(), it->OutputDefs().end(), inputs) != it->OutputDefs().end()) {
                   OrtValueIndex input_arg_idx;
                   ORT_THROW_IF_ERROR(ort_value_name_idx_map_.GetIdx(inputs->Name(), input_arg_idx));
                   auto& consumer_device = plan_.allocation_plan[input_arg_idx].location.device;
-                  if (devices_to_wait_on.find(consumer_device.Type()) == devices_to_wait_on.end()) {
+
+                  // launch WaitOnEPStep only when (same condition when initializing node_to_notification)
+                  // 1. input node and current node are in different streams, Or
+                  // 2. current node consumes a CPU tensor from an non-shape op.
+                  if (node_stream_map_[it->Index()] != i || consumer_device.Type() == OrtDevice::CPU) {
```
Comment (souptc) [inline onnxruntime/core/framework/allocation_planner.cc:1904]: the 2nd condition seems too loose. The real thing we need to catch in second condition is whether within the same stream is current node wait on a tensor copied from device. I think we can capture all the producer/consumer pairs of 2nd condition (cross device copy within same stream) and use it as a more clear condition here.

File: onnxruntime/core/framework/execution_provider.cc [hunk 12]
```diff
@@ -19,6 +19,7 @@ inline int MakeKey(int id, OrtMemType mem_type) {
 }  // namespace
 
 AllocatorPtr IExecutionProvider::GetAllocator(int device_id, OrtMemType mem_type) const {
+  ORT_ENFORCE(device_id == 0 || mem_type == OrtMemType::OrtMemTypeDefault, "It's illegal for non-CPU device to access CPU-related memory type");
```
Comment (souptc) [inline onnxruntime/core/framework/execution_provider.cc:22]: It seems not the check we want to do. I think the case we want to check is whether there is a call that search on OrtMemTypeDefault but the device_id doesn't match ExecutionProvider::GetDeviceId(). The main purpose is to check whether we have any place that the developer just hardcode the device_id to 0 on default memory type.

File: onnxruntime/test/framework/allocation_planner_test.cc [hunk 13]
```diff
@@ -71,6 +71,11 @@ struct UnaryNode {
 
   UnaryNode(onnxruntime::Graph& graph, onnxruntime::NodeArg* p_input_arg, onnxruntime::NodeArg* p_output_arg)
       : UnaryNode(graph, "Transpose", p_input_arg, p_output_arg) {}
+
+  UnaryNode(onnxruntime::Graph& graph, std::string& node_name, const std::string& op, std::vector<onnxruntime::NodeArg*>& inputs,
```
Comment (souptc) [inline onnxruntime/test/framework/allocation_planner_test.cc:75]: use std::string_view

File: onnxruntime/core/framework/allocation_planner.cc [hunk 14]
```diff
@@ -1809,40 +1808,44 @@ class PlannerImpl {
...
+                      && wait_handle != nullptr) {
+                    if (node_to_notification.find(node_index) == node_to_notification.end()) {
+                      node_to_notification[node_index] = plan_.notification_owners.size();
+                      plan_.notification_owners.push_back(i);
+                    }
+
+                    if (node_to_wait.find(it->Index()) == node_to_wait.end()) {
+                      node_to_wait[it->Index()] = std::vector<std::pair<WaitNotificationFn, NotificationIndex>>{std::make_pair(wait_handle, node_to_notification[node_index])};
+                    } else {
+                      node_to_wait[it->Index()].push_back(std::make_pair(wait_handle, node_to_notification[node_index]));
```
Comment (souptc) [inline onnxruntime/core/framework/allocation_planner.cc:1837]: you can unified it with insert.

Comment (souptc) [inline onnxruntime/core/framework/allocation_planner.cc:1837]: if node B consume multiple outputs of node A, will you insert multiple wait handles to the same notification into this list?

Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1837]: It will, as long as the wait_handle is valid under this case

Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc:1837]: only 1 wait step is needed for this case

Comment (jslhcl) [reply onnxruntime/core/framework/allocation_planner.cc]: for if (node_to_wait.find(it->Index()) == node_to_wait.end()) case, we cannot use insert() as there is no map instance under the key it->Index()

**Review by souptc — APPROVED** [review]

**Pull request merged**